File: bridging/BiquadFilterNode.cpp
#include "dart_api/dart_api.h"
#include "LabSound/LabSound.h"
#include "KeepNode.cpp"
#include "struct.h"
using namespace lab;

DART_EXPORT int createBiquadFilterNode(AudioContext* context) {
    auto sample = std::make_shared<BiquadFilterNode>(*context);
    return keepNode(sample);
}

DART_EXPORT int BiquadFilterNode_type(int nodeIndex) {
    return static_cast<int>(std::dynamic_pointer_cast<BiquadFilterNode>(audioNodes.find(nodeIndex)->second)->type());
}

DART_EXPORT void BiquadFilterNode_setType(int nodeIndex, int type) {
    std::dynamic_pointer_cast<BiquadFilterNode>(audioNodes.find(nodeIndex)->second)->setType(FilterType(type));
}

DART_EXPORT int BiquadFilterNode_frequency(int nodeId) {
    return keepAudioParam(nodeId, 1, std::static_pointer_cast<BiquadFilterNode>(audioNodes.find(nodeId)->second)->frequency());
}

DART_EXPORT int BiquadFilterNode_q(int nodeId) {
    return keepAudioParam(nodeId, 2, std::static_pointer_cast<BiquadFilterNode>(audioNodes.find(nodeId)->second)->q());
}

DART_EXPORT int BiquadFilterNode_gain(int nodeId) {
    return keepAudioParam(nodeId, 3, std::static_pointer_cast<BiquadFilterNode>(audioNodes.find(nodeId)->second)->gain());
}

DART_EXPORT int BiquadFilterNode_detune(int nodeId) {
    return keepAudioParam(nodeId, 4, std::static_pointer_cast<BiquadFilterNode>(audioNodes.find(nodeId)->second)->detune());
}

File: lab_sound_ffi/lib/src/lab_sound/biquad_filter_node.dart
import 'audio_context.dart';
import 'audio_node.dart';
import 'audio_param.dart';
import 'lab_sound.dart';

enum FilterType {
  filterNone,
  lowPass,
  highPass,
  bandpass,
  lowShelf,
  highShelf,
  peaking,
  notch,
  allPass,
}

class BiquadFilterNode extends AudioNode {

  late AudioParam frequency;
  late AudioParam q;
  late AudioParam gain;
  late AudioParam detune;

  BiquadFilterNode(AudioContext ctx)
      : super(ctx, LabSound().createBiquadFilterNode(ctx.pointer)) {
    frequency = AudioParam(this.ctx, nodeId, LabSound().BiquadFilterNode_frequency(nodeId));
    q = AudioParam(this.ctx, nodeId, LabSound().BiquadFilterNode_q(nodeId));
    gain = AudioParam(this.ctx, nodeId, LabSound().BiquadFilterNode_gain(nodeId));
    detune = AudioParam(this.ctx, nodeId, LabSound().BiquadFilterNode_detune(nodeId));
  }

  FilterType get type => FilterType.values[LabSound().BiquadFilterNode_type(nodeId)];
  set type(FilterType type) {
    LabSound().BiquadFilterNode_setType(nodeId, type.index);
  }

}
